pkg/pty/pty_linux.go

```diff
@@ -0,0 +1,85 @@
+// Copyright 2015-2017 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package pty provides basic pty support.
+// It implments much of exec.Command
+// but the Start() function starts two goroutines that relay the
+// data for Stdin, Stdout, and Stdout such that proper kernel pty
+// processing is done. We did not simply embed an exec.Command
+// as we can no guarantee that we can implement all aspects of it
+// for all time to come.
+package pty
+
+import (
+	"os"
+	"os/exec"
+	"strconv"
+	"syscall"
+	"unsafe"
+
+	"github.com/u-root/u-root/pkg/termios"
+)
+
+// pty support. We used to import github.com/kr/pty but what we need is not that complex.
+// Thanks to keith rarick for these functions.
+func New(cmd string, args ...string) (*Pty, error) {
+	tty, err := termios.New()
+	if err != nil {
+		return nil, err
+	}
+	restorer, err := tty.Get()
+	if err != nil {
+		return nil, err
+	}
+	ptm, err := os.OpenFile("/dev/ptmx", os.O_RDWR, 0)
+	if err != nil {
+		return nil, err
+	}
+
+	err = ptsunlock(ptm)
+	if err != nil {
+		return nil, err
+	}
+
+	sname, err := ptsname(ptm)
+	if err != nil {
+		return nil, err
+	}
+
+	// It can take a non-zero time for a pts to appear, it seems.
+	// Ten tries is reported to be far more than enough.
+	for i := 0; i < 10; i++ {
+		_, err := os.Stat(sname)
+		if err == nil {
+			break
+		}
+	}
+	pts, err := os.OpenFile(sname, os.O_RDWR|syscall.O_NOCTTY, 0)
```

Comment:
Go 1.9 might throw a wrench in this and a whole bunch of other open files in u-root (/dev/ptmx, esp the tty stuff, etc)

In Go 1.9, os.Files are added to an internal poll in the runtime and a call to Read and Write does not directly correspond to a call to sys_read(2) and sys_write(2) anymore. We may have to come up with our own file type that makes direct calls to syscall.Read and syscall.Write and uses syscall.Open at some point. (It's really dumb, and the Go people are breaking backwards compatibility here imho - you can't use os.File for named pipes anymore either.)